android/src/main/java/com/oney/WebRTCModule/UVCCamera2Enumerator.java

```diff
@@ -0,0 +1,107 @@
+package com.oney.WebRTCModule;
+
+import android.content.Context;
+import android.hardware.usb.UsbDevice;
+import android.util.Log;
+
+import androidx.annotation.Nullable;
+
+import com.jiangdg.ausbc.camera.CameraUvcStrategy;
+import com.jiangdg.ausbc.camera.bean.PreviewSize;
+
+import org.webrtc.Camera2Enumerator;
+import org.webrtc.CameraEnumerationAndroid;
+import org.webrtc.CameraVideoCapturer;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+
+/**
+ * Camera2Enumerator with UVC Camera support.
+ * <p/>
+ * This implementation only adds support for UVC camera devices.
+ * Ie. Device's (back and/or front) cameras will use the basic implementation of {@link Camera2Enumerator}.
+ */
+public class UVCCamera2Enumerator extends Camera2Enumerator  {
+
+    /** UVC camera names will be prefix with this value. Currently, there is no other way to
+     * easily distinguish between device's own and external uvc cameras in JS-side. */
+    public static final String UVC_PREFIX = "uvc-camera:";
+    private final Context context;
+
+    public UVCCamera2Enumerator(Context context) {
+        super(context);
+        this.context = context;
+    }
+
+    @Override
+    public String[] getDeviceNames() {
+        ArrayList<String> devicesNames = new ArrayList<>(Arrays.asList(super.getDeviceNames()));
+
+        try {
+            CameraUvcStrategy uvcStrategy = new CameraUvcStrategy(context);
+            List<UsbDevice> uvcCameras = uvcStrategy.getUsbDeviceList(null); // null := no filtering
+            if (uvcCameras != null) {
+                uvcCameras.forEach(camera -> devicesNames.add(UVC_PREFIX + camera.getDeviceName()));
+            }
+            uvcStrategy.unRegister();
+        } catch (Exception e) {
+            Log.e(WebRTCModule.TAG, "Error in uvc camera enumeration", e);
+        }

```

Comment:
Okay we fixed issues that @cristiano-linvix was having.

We changed the enumeration to use CameraUvcStrategy, which detects uvc cameras more robustly than my own implementation. Also, it now provides all available UVC devices instead of just only one.

@saghul

Reply:
Does it have any permissions implications?

Reply:
I was asked for camera permission when I started streaming to webrtc.

No other permissions were requested.

My `AndroidManifest.xml`

```xml
    <!-- react-native-webrtc -->
    <uses-feature android:name="android.hardware.camera" />
    <uses-feature android:name="android.hardware.camera.autofocus" />
    <uses-feature android:name="android.hardware.audio.output" />
    <uses-feature android:name="android.hardware.microphone" />

    <!-- react-native-webrtc -->
    <uses-permission android:name="android.permission.CAMERA" />
    <uses-permission android:name="android.permission.RECORD_AUDIO" />
    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
    <uses-permission android:name="android.permission.CHANGE_NETWORK_STATE" />
    <uses-permission android:name="android.permission.MODIFY_AUDIO_SETTINGS" />
```

Reply:
Will need to add something to the docs/handbook for this.
```xml
<uses-feature android:name="android.hardware.usb.host" />
```
But all good 👍🏻